src/python/pants/core/goals/generate_lockfiles.py

```diff
@@ -56,6 +58,48 @@ class ToolLockfileSentinel:
     options_scope: ClassVar[str]
 
 
+class UserLockfileRequests(Collection[LockfileRequest]):
+    """All user resolves for a particular language ecosystem to build.
+
+    Each language ecosystem should set up a subclass of `RequestedUserResolveNames` (see its
+    docstring), and implement a rule going from that subclass -> UserLockfileRequests. Each element
+    in the returned `UserLockfileRequests` should be a subclass of `LockfileRequest`, like
+    `PythonLockfileRequest`.
+    """
+
+
```

Comment:
I'll be honest: I don't fully understand why two `@union`s are necessary... but I suppose that it's because some bit of logic must be async, and so needs to be provided by a `@rule` rather than a method of the union class?

Reply:
It's because we need to first compute the entire universe of known resolves (the first union), then validate for ambiguity + filter to what was specified with --resolve, and then finally convert what was specified into actual LockfileRequest objects.